################################################################################
##        PAPER : Thambyrajah et al. 2024 Nat Comm
##                "Cis inhibition of NOTCH1 through JAGGED1 sustains embryonic Hematopoietic stem cell fate"
##  Description : This script is for plotting REVIGO* results
##       Author : Mar√≠a Maqueda (mmaqueda@researchmar.net) at BigSpinLab
################################################################################

# *Supek F et al. "REVIGO summarizes and visualizes long lists of Gene Ontology
# terms" PLoS ONE 2011. doi:10.1371/journal.pone.0021800


# NOTES:
# For generating this plot, it is required to import .tsv REVIGO results
# This code has been adapted from the automatically generated by REVIGO

################################################################################
## 1. Load packages
################################################################################

library( ggplot2 )
require(ggrepel)
library( scales )

################################################################################
## 2. Import results from REVIGO
################################################################################

res <- read.table(file = "Fig6E_Revigo_GOBPterms_results.tsv",
                  header=TRUE)

# Remove those terms that do not get to plot in the semantic space
res <- res [(res$PC_0 != "null" & res$PC_1 != "null"), ];

# Coordinates into numeric
res$PC_0 <- as.numeric( as.character(res$PC_0) );
res$PC_1 <- as.numeric( as.character(res$PC_1) );

################################################################################
## 3. Define the GO BP terms to highlight in the plot
################################################################################

highlight_GO <- c("TORC1 signaling", 
                  "chromatin remodeling",
                  "lymphocyte differentiation",
                  "myeloid cell differentiation",
                  "mitotic cell cycle phase transition",
                  "autophagy", 
                  "endosomal transport",
                  "chromosome separation",
                  "nuclear division",
                  "histone modification")

################################################################################
## 4. Plot 
################################################################################

# Basic plot
p1 <- ggplot( data = res );
p1 <- p1 + geom_point( aes( PC_0, PC_1, colour = Value, size = LogSize), alpha = I(0.6) );
p1 <- p1 + scale_colour_gradientn( colours = c("blue", "green", "yellow", "red"), limits = c( min(res$Value), 0), name="log10(adj.p)");
p1 <- p1 + geom_point( aes(PC_0, PC_1, size = LogSize), shape = 21, fill = "transparent", colour = I (alpha ("black", 0) ));
p1 <- p1 + scale_size( range=c(1, 4),name = "log10(GO size)") + theme_bw(); # + scale_fill_gradientn(colours = heat_hcl(7), limits = c(-300, 0) );

one.x_range = max(res$PC_0) - min(res$PC_0);
one.y_range = max(res$PC_1) - min(res$PC_1);
p1 <- p1 + xlim(min(res$PC_0)-one.x_range/10,max(res$PC_0)+one.x_range/10);
p1 <- p1 + ylim(min(res$PC_1)-one.y_range/10,max(res$PC_1)+one.y_range/10);

p1 <- p1 + labs (y = "semantic space y", x = "semantic space x");

p1 <- p1 + theme(legend.key = element_blank(), 
                 legend.key.width = unit(0.3, 'cm'),
                 legend.title = element_text(size=7),
                 panel.grid.major = element_blank(), 
                 panel.grid.minor = element_blank()) ;

# The ones to highlight
ex <- res [which(res$Name %in% highlight_GO), ];
p1 <- p1 + 
  geom_text_repel(data = ex, aes(x=PC_0, y=PC_1,label = Name),size = 2.5,
                  box.padding = unit(0.6,"lines"),
                  point_padding = unit(0.3,"lines"),
                  force=35) 
  
  #geom_text( data = ex, aes(PC_0, PC_1, label = Name), colour = I(alpha("black", 0.85)), size = 3 );

pdf(file = "Fig6E_nascentRNAseq_REVIGO_DEGs_UP_FcJAG1_exclusive.pdf", width=3.7, height=3)
p1
dev.off()





